README
Homepage: 
http://oss.alea-soluciones.com/trac/wiki/BoscliOss

See doc directory


debian/changelog
boscli-oss-core (0.2.8-1) unstable; urgency=low

  * Prefer exact matches when trying to find a function for
    a boscli comand / line

 -- Eduardo Ferro <eferro@alea-soluciones.com>  Tue, 05 Jun 2012 15:24:26 +0200

boscli-oss-core (0.2.7-1) unstable; urgency=low

  * Import plugins using imp module

 -- Eduardo Ferro <eferro@alea-soluciones.com>  Mon, 06 Feb 2012 15:27:49 +0100

boscli-oss-core (0.2.6-2) unstable; urgency=low

  * Bugfixes
  * accept new characters as special characters... (deprecated validation)

 -- Eduardo Ferro <eferro@alea-soluciones.com>  Mon, 06 Feb 2012 13:27:19 +0100

boscli-oss-core (0.2.5-1) unstable; urgency=low

  * Bugfixes

 -- Eduardo Ferro <eferro@alea-soluciones.com>  Sat, 29 Oct 2011 01:43:57 +0200

boscli-oss-core (0.2.4-1) unstable; urgency=low

  * Bugfixes

 -- Eduardo Ferro <eferro@alea-soluciones.com>  Thu, 27 Oct 2011 18:31:10 +0200

boscli-oss-core (0.2.3-1ubuntu1) lucid; urgency=low

  * Upgraded to lucid

 -- Eduardo Ferro <eferro@alea-soluciones.com>  Sun, 30 May 2010 07:51:23 +0200

boscli-oss-core (0.2.3-1) unstable; urgency=low

  * Support for stackable contexts (Cisco like contexts)
  * Support for execute commands outside the current context

 -- Eduardo Ferro <eferro@alea-soluciones.com>  Wed, 30 Sep 2009 12:31:22 +0200

boscli-oss-core (0.2.2-1) unstable; urgency=low

  * support autocompletion / interactive help, even if the user use abbreviations

 -- Eduardo Ferro <eferro@alea-soluciones.com>  Thu, 10 Sep 2009 08:58:49 +0200

boscli-oss-core (0.2.1-1) unstable; urgency=low

  * Normal parameters for bcli functions

 -- Eduardo Ferro <eferro@alea-soluciones.com>  Mon, 31 Aug 2009 08:00:46 +0200

boscli-oss-core (0.2.0-2) unstable; urgency=low

  * Debian package rewrite (using python-support)

 -- Eduardo Ferro <eferro@alea-soluciones.com>  Sun, 30 Aug 2009 16:41:33 +0200

boscli-oss-core (0.2.0-1) unstable; urgency=low

  * Cleanups and code reestucture

 -- Eduardo Ferro <eferro@alea-soluciones.com>  Wed, 12 Aug 2009 03:49:28 +0200

boscli-oss-core (0.1.9-5) unstable; urgency=low

  * dummy script for force to use python 2.5

 -- Eduardo Ferro <eferro@alea-soluciones.com>  Wed, 27 May 2009 17:04:03 +0200

boscli-oss-core (0.1.9-3) unstable; urgency=low

  * Minor bugfix

 -- Eduardo Ferro <eferro@alea-soluciones.com>  Fri, 15 May 2009 21:57:20 +0200

boscli-oss-core (0.1.9-2) unstable; urgency=low

  * Minor cleanup

 -- Eduardo Ferro <eferro@alea-soluciones.com>  Sun, 10 May 2009 22:09:16 +0200

boscli-oss-core (0.1.9-1) unstable; urgency=low

  * Initial release

 -- Eduardo Ferro <eferro@alea-soluciones.com>  Tue, 05 May 2009 08:20:05 +0200



src/bostypes.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# (c) Alea-Soluciones (Bifer) 2007, 2008, 2009
# Licensed under GPL v3 or later, see COPYING for the whole text

'''
  $Id:$
  $URL:$
  Alea-Soluciones (Bifer) (c) 2007
  Created: eferro - 22/7/2007
'''    


import socket
import re
import bosip
import os
import boscliutils
import glob
    
class BiferShellTypeManager:
    def __init__(self):
        self.types = {
            'STRING': BiferShellString(),
            'BOOL': BiferShellBool(),
            'STATE': BiferShellState(),
            'INT': BiferShellInt(),
            'FLOAT': BiferShellFloat(),
            'NET': BiferShellNet(),
            'NETMASK': BiferShellNetMask(),
            'NETIF': BiferShellNetIf(),
            'MAC': BiferShellMac(),
            'IP': BiferShellIp(),
            'CIDR': BiferShellCidr(),
            'HOST': BiferShellHost(),
            'IPHOST': BiferShellIpHost(),
            'WEBURL': BiferShellWebUrl(),
            'FILE': BiferShellFile(),
            'CMD': BiferShellCmd(),
            'KERNELVAR': BiferShellKernelVar(),
            'HEXNUMBER': BiferShellHexNumber(),
        }
    
    def add_type(self, type_name, type_class_instance):
        self.types[type_name] = type_class_instance

    def get_types(self):
        return self.types.keys()
        
    def get_type(self, type_name):
        return self.types[type_name]
    
    def set_cli(self, cli):
        for type in self.types.keys():
            self.types[type].set_cli(cli)
            
    def validate_value(self, type, value):
        return self.types[type].validate_value(value)
    def normalize(self, type, value):
        return self.types[type].normalize(value)
    def values(self, type, incomplete_word):
        return self.types[type].values(incomplete_word)
    def is_type(self, word):
        return word in self.types.keys()         
    
class BiferShellType:
    def __init__(self, doc = "Doc to be defined"):
        self.cli = None
        self.doc = doc

    def set_cli(self, cli):
        self.cli = cli
        
    def validate_value(self, value):
        res = value in self.values('')
        return res
    
    def normalize(self, value):
        return value
    
    def values(self, incomplete_word):
        return []

    def help(self):
        return self.doc
    
class BiferShellString(BiferShellType):
    def __init__(self): 
        BiferShellType.__init__(self, "Sequence of letters, digits or '_'. Ex: Xema")
    
    def validate_value(self, value):
        return True
    
    def normalize(self, value):
        return value.strip()
    
class BiferShellIp(BiferShellType):
    def __init__(self): 
        BiferShellType.__init__(self, "IP Address in dot-decimal notation: Ex: 192.168.1.100")
    def validate_value(self, value):
        try:
            if len(value.split('.')) != 4: return False

            mm = socket.inet_aton(value)
            return True # We got through that call without an error, so it is valid
        except socket.error:
            return False # There was an error, so it is invalid

class BiferShellHost(BiferShellType): 
    def __init__(self): 
        BiferShellType.__init__(self, "Hostname. Ex: www.google.com, bif-shp1")
    def validate_value(self, value):
        return True
    
class BiferShellIpHost(BiferShellType):
    def __init__(self):
        BiferShellType.__init__(self, "Hostname or IP Address. Ex: 192.168.10.200, www.google.com")
        self.ip_type = BiferShellIp()
        self.host_type = BiferShellHost()
    def validate_value(self, value):
        return self.ip_type.validate_value(value) or \
                self.host_type.validate_value(value)


class BiferShellWebUrl(BiferShellType):
    def __init__(self):
        BiferShellType.__init__(self, "Web URL, Ex: http://www.debian.org or https://www-nasa.gov")

    def validate_value(self, value):
        if value.startswith("http://") or value.startswith("https://"):
            return True
        else:
            return False

    def values(self, incomplete_word):
        if incomplete_word.startswith("http://") or incomplete_word.startswith("https://"):
            return []
        else:
            return ["http://", "https://"]
    
                
class BiferShellCidr(BiferShellType):
    def __init__(self): 
        BiferShellType.__init__(self,  
                                "Address in Classless Inter-Domain Routing  notation: Ex: 192.168.1.100/24")
    def validate_value(self, value):
        try:
            ip,bits = value.split('/')
            num_bits = int(bits)
            return True    
        except:
            return False
                
class BiferShellMac(BiferShellType):
    def __init__(self): 
        BiferShellType.__init__(self, "MAC or Ethernet Address in hex notation. Ex: 00:01:02:03:04:08")
    def validate_value(self, value):
        regexp = r"([0-9A-F][0-9A-F]:){5}([0-9A-F][0-9A-F])"
        match = re.match(regexp, value.upper())
        if match == None:
            return False
        else:
            return match.group().upper() == value.upper()


class BiferShellNet(BiferShellType): 
    def __init__(self): 
        BiferShellType.__init__(self)
    
class BiferShellNetMask(BiferShellType):
    def __init__(self): 
        BiferShellType.__init__(self, "Netmask in Dot-decimal Address. Ex: 255.255.255.0")

    def validate_value(self, value):
        return  bosip.Ip.validate_netmask(value)
    

class BiferShellNetIf(BiferShellType):
    def __init__(self): 
        BiferShellType.__init__(self, "Ethernet interface. Ex: eth0, eth1, eth0:0")
    def validate_value(self, value):
        return value in self.values('')
        
    def normalize(self, value):
        return value
    
    def values(self, incomplete_word):
        return ['eth0','eth0:0', 'eth1', 'eth1:0', 'eth2','eth2:0', 'br0']

class BiferShellInt(BiferShellType):
    def __init__(self): 
        BiferShellType.__init__(self, "Integer value. Ex: 42")
    def validate_value(self, value):
        try:
            int(value)
            return True
        except:
            return False
        
class BiferShellFloat(BiferShellType):
    def __init__(self): 
        BiferShellType.__init__(self, "Decimal number")
    def validate_value(self, value):
        try:
            float(value)
            return True
        except:
            return False


class BiferShellBool(BiferShellType):
    def __init__(self): 
        BiferShellType.__init__(self, "Coditional value. Accepted values 'true', 'false', '1', '0'")
    def validate_value(self, value):
        return value in self.values('')
        
    def normalize(self, value):
        return value
    
    def values(self, incomplete_word):
        return ['true', 'false', '1', '0']

class BiferShellState(BiferShellType):
    def __init__(self): 
        BiferShellType.__init__(self, "State value. Accepted values 'on', 'off', 'up', 'down'")
    def validate_value(self, value):
        return value in self.values('')
        
    def normalize(self, value):
        return value
    
    def values(self, incomplete_word):
        return ['on', 'off', 'up', 'down']
    
class BiferShellFile(BiferShellType):
    def __init__(self): 
        BiferShellType.__init__(self, "Filename with absolute path")
    def validate_value(self, Value):
        # TODO: Implement
        return True
        
    def normalize(self, value):
        return value

    def files(self, incomplete_path):
        return  [f for f in glob.glob(incomplete_path + '*') if f[-1:] not in ['~', '#']]
    
    def values(self, incomplete_word):
        paths = []
        for f in self.files(incomplete_word):
            if os.path.isdir(f):
                paths.append(f + '/')
            else:
                paths.append(f)

        if len(paths) == 1 and os.path.isdir(paths[0]):
            return self.values(paths[0])
        return paths
    

class BiferShellKernelVar(BiferShellType):
    def __init__(self):
        BiferShellType.__init__(self, "Kernel variable. Ex: net.unix.max_dgram_qlen or fs.mqueue.msgsize_max")
        self.kernel_vars=[]
        for line in os.popen("sysctl -A 2>/dev/null").readlines():
            self.kernel_vars.append(line.split('=')[0].strip())
    
    def values(self, incomplete_word):
        return self.kernel_vars
    
class BiferShellCmd(BiferShellType):
    def __init__(self): 
        BiferShellType.__init__(self, "Cli available functions")
        
    def values(self, incomplete_word):
        if self.cli == None:
            return []
        else:
            functions = self.cli.get_functions()
            cmds = sorted(list(set([a.split('_')[1] for a in functions])))
            return cmds


 
 
class BiferShellHexNumber(BiferShellType):
    def __init__(self): 
        BiferShellType.__init__(self, "Hex number value. Ex: FF02FE3A")
    def validate_value(self, value):
        try:
            int(value, 16)
            return True
        except:
            return False
 



class BiferGenericType(BiferShellType):
    def __init__(self, values_func, short_help = None):
        self.__values_func = values_func
        if short_help != None:
            doc = short_help
        else:
            doc = "Doc to be defined"
            if self.__values_func.__doc__ != None:
                doc = self.__values_func.__doc__
        BiferShellType.__init__(self, doc)
        
    def validate_value(self, value):
        return value in self.values('')

    def values(self, incomplete_word):
        return self.__values_func(incomplete_word)


class BiferOptionsType(BiferShellType):
    def __init__(self, options_list, doc = None):
        self.__options = options_list
        if doc == None:
            doc = "Options %s" % str(self.__options)

        BiferShellType.__init__(self, doc)
        
    def validate_value(self, value):
        return value in self.values('')

    def values(self, incomplete_word):
        return self.__options




def test1(incomplete_word):
    """Type documentation"""
    print "test1"
    print incomplete_word    
    return ['one', 'two']

if __name__ == '__main__':
    type1 = BiferGenericType(test1)
    print type1.values('incomp')
    print type1.help()

    type2 = BiferOptionsType(['Op1', 'Op2', 'Op3'])
    print type2.values('incomp')
    print type2.help()

    type3 = BiferGenericType(lambda x: ['Op1', 'Op2'])
    print type3.values('incomp')
    print type3.help()

    type4 = BiferOptionsType(['Op1', 'Op2', 'Op3'], "Options type example")
    print type4.values('incomp')
    print type4.help()
    

    type5 = BiferGenericType(test1, "with short help")
    print type5.values('incomp')
    print type5.help()
